app/gui/src/project-view/stores/project/index.ts

```diff
@@ -327,6 +328,86 @@ export function createProjectStore(
     })
   }
 
+  // Maximum number of in-progress expressions.
+  const MAX_IN_PROGRESS = 5
+
+  const inProgress = ref(0)
+  const queueLength = ref(0)
+
+  function queuedExecuteExpression(
+    expressionId: ExternalId,
+    expression: string,
+    timeoutMs: number = 5000,
+  ): Promise<Result<any> | null> {
+    if (inProgress.value > MAX_IN_PROGRESS) {
+      if (timeoutMs < 0) {
+        return Promise.reject(Err(`queuedExecuteExpression: Execution timed out.`))
+      }
+
+      queueLength.value += 1
+      const pause = queueLength.value * 250
+      return new Promise((resolve) => setTimeout(resolve, pause)).then(() => {
+        queueLength.value -= 1
```

Comment:
In this queue design, requests may wait longer than necessary; they are also executed out of order, potentially starving some requests. We have a `ConditionVariable` helper we currently use for efficient concurrency control in the `queryClient`.